### PR TITLE
[WIN32SS][NTGDI] Strictly check in XFORMOBJ_iSetXform

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
JIRA issue: [CORE-15554](https://jira.reactos.org/browse/CORE-15554)
Check if the determinant of the given transformation matrix is non-zero. If zero, the function returns `DDI_ERROR` without change.

### CODE REVIEW

**Comment by katahiromz** [comment]
Any better ideas?

**Comment by HBelusca** [comment]
This looks similar to what you are doing: https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/xformobj.c;hb=d182f449fc284bfd263f5b84b87b0e311869479c#l238

**Comment by katahiromz** [comment]
I think this part needs severely optimized. Do you think this code is optimized?

**Comment by katahiromz** [comment]
There were no objection. Going to the testing stage...

**Comment by HBelusca** [comment]
> I think this part needs severely optimized. Do you think this code is optimized?

If you see that execution is much slower when using one of the proposed versions, then use the faster one. Otherwise if both run with the same time, prefer using the existing function, **unless** it requires to completely re-create a new "matrix" object...

**Comment by tkreuzer** [comment]
> 
> 
> I think this part needs severely optimized. Do you think this code is optimized?

Yes, that part is optimized. The only thing that can be more optimized is replacing the subtraction with a comparison for equality, but for that I'd like to see a mathmatical proof, that it's exactly identical to the result of the subtraction being 0 (think of 0.0 vs -0.0, denormals).

File: win32ss/gdi/ntgdi/xformobj.c [hunk 1]
```diff
@@ -166,8 +169,27 @@ XFORMOBJ_iSetXform(
...
+    ef1 = pmx->efM11;
+    ef2 = pmx->efM22;
+    ef3 = pmx->efM12;
+    ef4 = pmx->efM21;
+
     /* Update accelerators and return complexity */
-    return XFORMOBJ_UpdateAccel(pxo);
+    Hint = XFORMOBJ_UpdateAccel(pxo);
+
+    if (Hint == GX_SCALE || Hint == GX_GENERAL)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/xformobj.c]: Thanks, that's smart.

File: win32ss/gdi/ntgdi/xformobj.c [hunk 2]
```diff
@@ -151,12 +151,15 @@ XFORMOBJ_iSetXform(
...
+    ULONG Hint;
+    FLOATOBJ ef1, ef2, ef3, ef4;
 
     /* Check parameters */
     if (!pxo || !pxform) return DDI_ERROR;
 
-    /* Check if the xform is valid */
-    if ((pxform->eM11 == 0) || (pxform->eM22 == 0)) return DDI_ERROR;
+    /* Save */
+    mxSave = *pmx;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/xformobj.c]: OK.

**Comment by katahiromz** [comment]
Done.

**Pull request merged**